File: tests/test_apps/php/smoke_test.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use TreeSitterLanguagePack\LanguagePack;

final class SmokeTest extends TestCase
{
    private static string $fixturesDir;
    private LanguagePack $pack;

    public static function setUpBeforeClass(): void
    {
        self::$fixturesDir = dirname(__DIR__) . '/fixtures';
        // Download required languages
        LanguagePack::download(['python', 'javascript', 'rust', 'go', 'ruby', 'java', 'c', 'cpp']);
    }

    protected function setUp(): void
    {
        $this->pack = new LanguagePack();
    }

    private static function loadFixtures(string $name): array
    {
        $json = file_get_contents(self::$fixturesDir . '/' . $name);
        return json_decode($json, true, 512, JSON_THROW_ON_ERROR);
    }

    /**
     * @dataProvider basicFixtureProvider
     */
    public function testBasicFixture(array $fixture): void
    {
        match ($fixture['test']) {
            'language_count' => $this->assertGreaterThanOrEqual(
                $fixture['expected_min'],
                $this->pack->languageCount(),
                "language_count < expected min {$fixture['expected_min']}"
            ),
            'has_language' => $this->assertSame(
                $fixture['expected'],
                $this->pack->hasLanguage($fixture['language']),
                "has_language({$fixture['language']})"
            ),
            'available_languages' => $this->assertAvailableLanguagesContains(
                $fixture['expected_contains']
            ),
            default => $this->fail("Unknown test type: {$fixture['test']}"),
        };
    }

    public static function basicFixtureProvider(): iterable
    {
        $fixtures = self::loadFixtures('basic.json');
        foreach ($fixtures as $fixture) {
            yield $fixture['name'] => [$fixture];
        }
    }

    /**
     * @dataProvider processFixtureProvider
     */
    public function testProcessFixture(array $fixture): void
    {
        $source = $fixture['source'];
        $config = json_encode($fixture['config']);
        $result = $this->pack->process($source, $config);
        $expected = $fixture['expected'];

        if (isset($expected['language'])) {
            $this->assertSame($expected['language'], $result['language']);
        }
        if (isset($expected['structure_min'])) {
            $structureCount = count($result['structure'] ?? []);
            $this->assertGreaterThanOrEqual($expected['structure_min'], $structureCount,
                "structure count {$structureCount} < min {$expected['structure_min']}");
        }
        if (isset($expected['imports_min'])) {
            $importsCount = count($result['imports'] ?? []);
            $this->assertGreaterThanOrEqual($expected['imports_min'], $importsCount,
                "imports count {$importsCount} < min {$expected['imports_min']}");
        }
        if (isset($expected['error_count'])) {
            $errorCount = $result['metrics']['error_count'] ?? 0;
            $this->assertSame($expected['error_count'], $errorCount);
        }
        if (isset($expected['metrics_total_lines_min'])) {
            $totalLines = $result['metrics']['total_lines'] ?? 0;
            $this->assertGreaterThanOrEqual($expected['metrics_total_lines_min'], $totalLines,
                "total_lines {$totalLines} < min {$expected['metrics_total_lines_min']}");
        }
    }

    public static function processFixtureProvider(): iterable
    {
        $fixtures = self::loadFixtures('process.json');
        foreach ($fixtures as $fixture) {
            yield $fixture['name'] => [$fixture];
        }
    }

    /**
     * @dataProvider chunkingFixtureProvider
     */
    public function testChunkingFixture(array $fixture): void
    {
        $source = $fixture['source'];
        $config = json_encode($fixture['config']);
        $result = $this->pack->process($source, $config);
        $expected = $fixture['expected'];

        if (isset($expected['chunks_min'])) {
            $chunksCount = count($result['chunks'] ?? []);
            $this->assertGreaterThanOrEqual($expected['chunks_min'], $chunksCount,
                "chunks count {$chunksCount} < min {$expected['chunks_min']}");
        }
    }

    public static function chunkingFixtureProvider(): iterable
    {
        $fixtures = self::loadFixtures('chunking.json');
        foreach ($fixtures as $fixture) {
            yield $fixture['name'] => [$fixture];
        }
    }

    private function assertAvailableLanguagesContains(array $expected): void
    {
        $langs = $this->pack->availableLanguages();
        foreach ($expected as $lang) {
            $this->assertContains($lang, $langs, "available_languages missing '{$lang}'");
        }
    }

    // Download API tests
    public function testDownloadedLanguagesReturnsArray(): void
    {
        $langs = LanguagePack::downloadedLanguages();
        $this->assertIsArray($langs);
    }

    public function testManifestLanguagesReturnsArrayWith50Plus(): void
    {
        $langs = LanguagePack::manifestLanguages();
        $this->assertIsArray($langs);
        $this->assertGreaterThan(50, count($langs), "manifestLanguages should return 50+ languages");
    }

    public function testCacheDirReturnsNonEmptyString(): void
    {
        $dir = LanguagePack::cacheDir();
        $this->assertIsString($dir);
        $this->assertGreaterThan(0, strlen($dir), "cacheDir should return non-empty string");
    }

    public function testInitDoesNotThrow(): void
    {
        LanguagePack::init();
        $this->assertTrue(true);
    }

    // Parse validation tests
    public function testParsesPythonCode(): void
    {
        $tree = $this->pack->parseString('python', "def hello(): pass\n");
        $this->assertSame('module', $tree->rootNodeType());
        $this->assertGreaterThanOrEqual(1, $tree->rootChildCount());
        $this->assertFalse($tree->hasErrorNodes());
    }

    public function testErrorsOnInvalidLanguage(): void
    {
        $this->expectException(\RuntimeException::class);
        $this->pack->parseString('nonexistent_xyz_123', 'code');
    }

    public function testHasLanguageReturnsFalseForNonexistent(): void
    {
        $result = $this->pack->hasLanguage('nonexistent_xyz_123');
        $this->assertFalse($result);
    }
}

File: tests/test_apps/csharp/SmokeTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using TreeSitterLanguagePack;
using Xunit;

namespace TestApp;

public class SmokeTest
{
    private static readonly string FixturesDir = Path.Combine(
        Directory.GetCurrentDirectory(), "..", "..", "..", "..", "fixtures");

    private static List<JsonElement> LoadFixtures(string name)
    {
        var json = File.ReadAllText(Path.Combine(FixturesDir, name));
        return JsonSerializer.Deserialize<List<JsonElement>>(json)!;
    }

    private static Dictionary<string, object?> JsonElementToDictionary(JsonElement element)
    {
        var dict = new Dictionary<string, object?>();
        if (element.ValueKind == JsonValueKind.Object)
        {
            foreach (var property in element.EnumerateObject())
            {
                dict[property.Name] = JsonElementToObject(property.Value);
            }
        }
        return dict;
    }

    private static object? JsonElementToObject(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.String => element.GetString(),
            JsonValueKind.Number => element.GetDouble(),
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Null => null,
            JsonValueKind.Object => JsonElementToDictionary(element),
            JsonValueKind.Array => element.EnumerateArray().Select(JsonElementToObject).ToList(),
            _ => null
        };
    }

    // Basic fixtures tests
    [Theory]
    [MemberData(nameof(BasicFixtures))]
    public void TestBasicFixture(Dictionary<string, object?> fixture)
    {
        var test = (string)fixture["test"]!;
        switch (test)
        {
            case "language_count":
                {
                    var count = TsPackClient.LanguageCount();
                    var expectedMin = Convert.ToInt32(fixture["expected_min"]);
                    Assert.True(count >= expectedMin,
                        $"language_count {count} < expected min {expectedMin}");
                    break;
                }
            case "has_language":
                {
                    var language = (string)fixture["language"]!;
                    var result = TsPackClient.HasLanguage(language);
                    var expected = (bool)fixture["expected"]!;
                    Assert.Equal(expected, result);
                    break;
                }
            case "available_languages":
                {
                    var langs = TsPackClient.AvailableLanguages();
                    var expectedContains = (List<object?>)fixture["expected_contains"]!;
                    foreach (var lang in expectedContains)
                    {
                        Assert.Contains((string)lang!, langs);
                    }
                    break;
                }
            default:
                throw new InvalidOperationException($"Unknown test type: {test}");
        }
    }

    public static IEnumerable<object[]> BasicFixtures()
    {
        var fixtures = LoadFixtures("basic.json");
        foreach (var fixture in fixtures)
        {
            var dict = JsonElementToDictionary(fixture);
            yield return new object[] { dict };
        }
    }

    // Process fixtures tests
    [Theory]
    [MemberData(nameof(ProcessFixtures))]
    public void TestProcessFixture(Dictionary<string, object?> fixture)
    {
        var source = (string)fixture["source"]!;
        var configMap = (Dictionary<string, object?>)fixture["config"]!;

        var config = new ProcessConfig
        {
            Language = (string)configMap["language"]!
        };

        if (configMap.TryGetValue("structure", out var structureVal))
        {
            config.Structure = (bool)structureVal!;
        }
        if (configMap.TryGetValue("imports", out var importsVal))
        {
            config.Imports = (bool)importsVal!;
        }
        if (configMap.TryGetValue("exports", out var exportsVal))
        {
            config.Exports = (bool)exportsVal!;
        }
        if (configMap.TryGetValue("comments", out var commentsVal))
        {
            config.Comments = (bool)commentsVal!;
        }
        if (configMap.TryGetValue("chunk_max_size", out var chunkVal))
        {
            config.ChunkMaxSize = Convert.ToInt32(chunkVal);
        }

        var expected = (Dictionary<string, object?>)fixture["expected"]!;
        var result = TsPackClient.Process(source, config);

        if (expected.TryGetValue("language", out var expectedLang))
        {
            Assert.Equal((string)expectedLang!, result.Language);
        }

        if (expected.TryGetValue("structure_min", out var structureMin))
        {
            var min = Convert.ToInt32(structureMin);
            Assert.True(result.Structure.Count >= min,
                $"structure count {result.Structure.Count} < min {min}");
        }

        if (expected.TryGetValue("imports_min", out var importsMin))
        {
            var min = Convert.ToInt32(importsMin);
            Assert.True(result.Imports.Count >= min,
                $"imports count {result.Imports.Count} < min {min}");
        }

        if (expected.TryGetValue("error_count", out var errorCount))
        {
            var expectedCount = Convert.ToInt32(errorCount);
            Assert.Equal(expectedCount, result.Metrics.ErrorCount);
        }

        if (expected.TryGetValue("metrics_total_lines_min", out var totalLinesMin))
        {
            var min = Convert.ToInt32(totalLinesMin);
            Assert.True(result.Metrics.TotalLines >= min,
                $"total_lines {result.Metrics.TotalLines} < min {min}");
        }
    }

    public static IEnumerable<object[]> ProcessFixtures()
    {
        var fixtures = LoadFixtures("process.json");
        foreach (var fixture in fixtures)
        {
            var dict = JsonElementToDictionary(fixture);
            yield return new object[] { dict };
        }
    }

    // Chunking fixtures tests
    [Theory]
    [MemberData(nameof(ChunkingFixtures))]
    public void TestChunkingFixture(Dictionary<string, object?> fixture)
    {
        var source = (string)fixture["source"]!;
        var configMap = (Dictionary<string, object?>)fixture["config"]!;
        var expected = (Dictionary<string, object?>)fixture["expected"]!;

        var config = new ProcessConfig
        {
            Language = (string)configMap["language"]!
        };

        if (configMap.TryGetValue("structure", out var structureVal))
        {
            config.Structure = (bool)structureVal!;
        }
        if (configMap.TryGetValue("chunk_max_size", out var chunkVal))
        {
            config.ChunkMaxSize = Convert.ToInt32(chunkVal);
        }

        var result = TsPackClient.Process(source, config);

        if (expected.TryGetValue("chunks_min", out var chunksMin))
        {
            var min = Convert.ToInt32(chunksMin);
            Assert.True(result.Chunks.Count >= min,
                $"chunks count {result.Chunks.Count} < min {min}");
        }
    }

    public static IEnumerable<object[]> ChunkingFixtures()
    {
        var fixtures = LoadFixtures("chunking.json");
        foreach (var fixture in fixtures)
        {
            var dict = JsonElementToDictionary(fixture);
            yield return new object[] { dict };
        }
    }

    // Parse validation tests
    [Fact]
    public void ParsesPythonCode()
    {
        using var tree = TsPackClient.Parse("python", "def hello(): pass\n");
        Assert.NotNull(tree);
        Assert.Equal("module", tree.RootNodeType());
        Assert.True(tree.RootChildCount() >= 1);
        Assert.False(tree.HasErrorNodes());
    }

    [Fact]
    public void ErrorsOnInvalidLanguage()
    {
        Assert.Throws<TsPackException>(() =>
            TsPackClient.Parse("nonexistent_xyz_123", "code"));
    }

    [Fact]
    public void HasLanguageReturnsFalseForNonexistent()
    {
        var result = TsPackClient.HasLanguage("nonexistent_xyz_123");
        Assert.False(result);
    }

    [Fact]
    public void AvailableLanguagesReturnsNonEmptyArray()
    {
        var langs = TsPackClient.AvailableLanguages();
        Assert.NotNull(langs);
        Assert.True(langs.Length > 0, "AvailableLanguages should return at least one language");
        Assert.Contains("python", langs);
        Assert.Contains("javascript", langs);
    }

    [Fact]
    public void LanguageCountIsPositive()
    {
        var count = TsPackClient.LanguageCount();
        Assert.True(count > 0, "LanguageCount should be positive");
    }

    [Fact]
    public void ParseTreeDisposesCleanly()
    {
        var tree = TsPackClient.Parse("python", "x = 1\n");
        tree.Dispose();
        // Second dispose should be a safe no-op
        tree.Dispose();
    }

    [Fact]
    public void ParseTreeHasErrorNodesForInvalidSyntax()
    {
        using var tree = TsPackClient.Parse("python", "def ():\n");
        Assert.True(tree.HasErrorNodes(), "Invalid Python syntax should produce error nodes");
        Assert.True(tree.ErrorCount() > 0, "ErrorCount should be positive for invalid syntax");
    }

    [Fact]
    public void ParseTreeContainsNodeType()
    {
        using var tree = TsPackClient.Parse("python", "def hello(): pass\n");
        Assert.True(tree.ContainsNodeType("function_definition"),
            "Python function should contain function_definition node");
    }

    [Fact]
    public void ParseTreeToSexpReturnsNonEmpty()
    {
        using var tree = TsPackClient.Parse("python", "x = 1\n");
        var sexp = tree.ToSexp();
        Assert.NotNull(sexp);
        Assert.True(sexp!.Length > 0, "S-expression should be non-empty");
    }
}
